Content.Server/Singularity/EntitySystems/GravityWellSystem.cs

```diff
@@ -61,201 +58,226 @@ public override void Update(float frameTime)
+            return;
+        
+        var physicsQuery = EntityManager.GetEntityQuery<PhysicsComponent>();
+        var xformQuery = EntityManager.GetEntityQuery<TransformComponent>();
+        UpdateBeforeSolve(uid, physicsQuery, xformQuery, gravWell, xform);
     }
 
     /// <summary>
-    /// Greates a gravitational pulse, shoving around all entities within some distance of an epicenter.
+    /// Updates the gravitational force applied to all objects captured by all gravity wells.
     /// </summary>
-    /// <param name="uid">The entity at the epicenter of the gravity pulse.</param>
-    /// <param name="maxRange">The maximum distance at which entities can be affected by the gravity pulse.</param>
-    /// <param name="minRange">The minimum distance at which entities can be affected by the gravity pulse.</param>
-    /// <param name="baseRadialDeltaV">The base radial velocity that will be added to entities within range towards the center of the gravitational pulse.</param>
-    /// <param name="baseTangentialDeltaV">The base tangential velocity that will be added to entities within countrclockwise around the center of the gravitational pulse.</param>
-    /// <param name="xform">(optional) The transform of the entity at the epicenter of the gravitational pulse.</param>
-    public void GravPulse(EntityUid uid, float maxRange, float minRange, float baseRadialDeltaV = 0.0f, float baseTangentialDeltaV = 0.0f, TransformComponent? xform = null)
+    /// <param name="uid">The uid of the gravity well to update.</param>
+    /// <param name="physicsQuery">A query for the physics body states of affected entities.</param>
+    /// <param name="xformQuery">A query for the position states of affected entities.</param>
+    /// <param name="gravWell">The state of the gravity well to update.</param>
+    /// <param name="xform">The position state of the gravity well.</param>
+    public void UpdateBeforeSolve(
+        EntityUid uid,
+        EntityQuery<PhysicsComponent> physicsQuery, EntityQuery<TransformComponent> xformQuery,
+        GravityWellComponent? gravWell = null, TransformComponent? xform = null)
     {
-        if (Resolve(uid, ref xform))
-            GravPulse(xform.Coordinates, maxRange, minRange, baseRadialDeltaV, baseTangentialDeltaV);
+        if(!Resolve(uid, ref gravWell, ref xform))
+            return;
+
+        var epicenter = xform.MapPosition;
+        var minRange2 = MathF.Max(gravWell.MinRange * gravWell.MinRange, MinGravWellRange);
+        var maxRange2 = gravWell.MaxRange * gravWell.MaxRange;
+        foreach(var entity in gravWell.Captured)
+        {
+            if(!physicsQuery.TryGetComponent(entity, out var physicsBody)
+            || !xformQuery.TryGetComponent(entity, out var entityXform))
+                continue;
+                    
+            var position = entityXform.MapPosition;
+            if (epicenter.MapId != position.MapId)
+                continue;
+            
```

Comment:
This should be re-arranged to be more optimal, checking mapid is free but getting worldposition is expensive for example.

Reply:
I've made some changes. It should be slightly faster now.